docs: start writing CRDs guide (#1116)
port-labs/port-docs: docs/guides-and-tutorials/manage-resources-using-k8s-crds.md

---
sidebar_position: 11
title: Manage resources using Kubernetes CRDs
---

import PortTooltip from "/src/components/tooltip/tooltip.jsx"

# Manage resources using Kubernetes CRDs

### Introduction

> Kubernetes provides a great way of extending its API which is via Custom Resource Definitions (CRDs). CRDs allow you to define your API objects and controllers to manage them. This is a powerful feature that allows you to extend Kubernetes to manage any kind of resources.

Reviewer: ```suggestion
Kubernetes provides a great way of extending its API via Custom Resource Definitions (CRDs). CRDs allow you to define your API objects and controllers to manage them. This is a powerful feature that allows you to extend Kubernetes to manage any kind of resource.
```

This guide will show you how to integrate Kubernetes CRDs with Port and expose them in the Port UI for developers to use.

:::tip Prerequisites
- Port account
- Kubernetes cluster

> - Deployed operator that listens to CRD changes

Reviewer: @danielsinai This can be unclear, perhaps add a relevant link

- GitHub account
:::

### The goal of this guide

> - Developers will be able to use any CRDs directly in a UI.

Reviewer: ```suggestion
- Developers will be able to use any CRDs directly in Port's UI.
```

- Platform engineers will be able to query the data and get insights about the usage of the CRDs and Kubernetes resources.
- Platform engineers will be able to serve any CRDs to developers in a self-service manner.

> In this guide, we will deploy Port's [Kubernetes Exporter](/build-your-software-catalog/sync-data-to-catalog/kubernetes/kubernetes.md) to export Kubernetes CRDs to Port as <PortTooltip id="blueprint">blueprints</PortTooltip> and the relevant <PortTooltip id="action">actions</PortTooltip> to create, update and delete those resources and listen to changes in those resources and reflect them in Port's UI without the need of creating a mapping and blueprints upfront.

Reviewer: This is a long and complicated sentence. Rephrase to make it clearer

<img src='/img/guides/visualizeCRDs.svg' width='100%' border='1px' />

> Then we will connect a GitHub account using Port's [GitHub integration](/build-your-software-catalog/sync-data-to-catalog/git/github/github.md) to provision CRs directly into the Kubernetes cluster, or optionally with GitOps.

Reviewer: ```suggestion
Next, we will connect a GitHub account using Port's [GitHub integration](/build-your-software-catalog/sync-data-to-catalog/git/github/github.md) to provision CRs directly into the Kubernetes cluster, or optionally with GitOps.
```

After completing it, you will get a sense of how it can benefit different personas in your organization:

<img src='/img/guides/createCRs.svg' width='100%' border='1px' />

> ### 1. (Optional) Creating a Crossplane XRD & Composition or using an existing CRD

Reviewer: ```suggestion
### 1. (Optional) Create a Crossplane XRD & Composition or use an existing CRD
```

**If you already have a CRD that you want to expose in Port, you can skip this step.**

If you don't have a CRD and you would like to create one, you can use Crossplane's XRD to create a CRD like we do in this guide or to install any operator that applies CRD into your cluster. You can follow the [Crossplane XRD documentation](https://docs.crossplane.io/latest/concepts/composite-resource-definitions/) to create a CRD.
For this guide, we followed the [AWS DynamoDB composition example](https://docs.crossplane.io/latest/getting-started/provider-aws-part-2/)

> ### 2. Installing the Kubernetes Exporter with `crdsToDiscover` flag

Reviewer: ```suggestion
### 2. Install the Kubernetes Exporter with the `crdsToDiscover` flag
```

The Kubernetes Exporter can be installed with the `crdsToDiscover` flag which is a JQ pattern to discover and export CRDs to Port as blueprints and actions. In this example we will use [Helm](https://helm.sh/) to install the Kubernetes Exporter, but for more installation options please visit the [Kubernetes Exporter documentation](/build-your-software-catalog/sync-data-to-catalog/kubernetes/kubernetes.md#installation)

> :::note

Reviewer: add a title to the admonition. Also, consider using a type other than "note", since it's not very visually appealing

```bash
# highlight-start
echo 'crdsToDiscover: ".metadata.ownerReferences[0].kind == \"CompositeResourceDefinition\" and .spec.scope != \"Namespaced\""' > config.yaml
# highlight-end 

helm upgrade --install my-port-k8s-exporter port-labs/port-k8s-exporter \
    --create-namespace --namespace port-k8s-exporter \
    --set secret.secrets.portClientId=YOUR_PORT_CLIENT_ID \
    --set secret.secrets.portClientSecret=YOUR_PORT_CLIENT_SECRET \
    --set stateKey="k8s-exporter"  \
    --set eventListenerType="POLLING"  \
    --set overwriteConfigurationOnRestart=true \
    --set "extraEnv[0].name"="CLUSTER_NAME" \
    --set "extraEnv[0].value"=YOUR_PORT_CLUSTER_NAME \
    --set-file configMap.config=config.yaml
```

After the Kubernetes Exporter is installed, you can check the logs to see if the CRDs are being discovered and exported to Port.
If everything succeeded you should see the relevant CRDs as `blueprints` and `actions`, and any existing CRs as `entities` in Port's catalog.

> ### 3. Connecting a GitHub workflow

Reviewer: ```suggestion
### 3. Connect a GitHub workflow
```

Now, to execute the action we need to add some workflow to it for Port to trigger the action, the workflow that we will configure will do the operations in the Kubernetes cluster using the `kubeconfig` that we will provide as a secret in the GitHub repository.

1. Go to the [GitHub control plane example repo](https://github.com/port-labs/control-plane-demo) and use it as a template to create your own repo by clicking on the `Use this template` button.

> ![CRD Action](/img/guides/useThisTemplate.png)

Reviewer: please use the `<img src="...">` tag to add images, like you did above. This allows easier control over the size and other params, and ensures consistency across the docs. Also, add a 1px border like you did above.

2. [Install Port's GitHub app](/build-your-software-catalog/sync-data-to-catalog/git/github/github.md#installation) in the GitHub account where you cloned the template.

3. [Add GitHub secrets](https://docs.github.com/en/actions/security-guides/using-secrets-in-github-actions#creating-secrets-for-a-repository) the following secrets:
    - `PORT_CLIENT_ID` - Your Port Client ID.
    - `PORT_CLIENT_SECRET` - Your Port Client Secret.
    - `KUBE_CONFIG` - The base64 `kubeconfig` of the Kubernetes cluster where the CRDs are deployed. You can get the kubeconfig by running the following command:
        Here is a snippet to encode your `kubeconfig` and copy it to your clipboard:
        ```bash
        cat $HOME/.kube/config | base64 | pbcopy
        ```
4. Edit the org and repo of Port's Action, to match yours you can do that by (make sure to do it to all actions create/delete/update):
    * Go to the [Self-Service Tab](https://app.getport.io/self-serve)
    * Hover the wanted action 

>     * Click on the 3 dots 

Reviewer: ```suggestion
    * Click on the `...` button
```

    * Click on `Edit Action` 
    * Click in the `Backend` tab there change the `organization` and `repository` fields to match yours.

    ![CRD Action](/img/guides/changeOrgRepo.png)

:::note

> **GitOps**

Reviewer: add a title to the admonition, perhaps just move the `GitOps` to be the title
